File: javascript/dnd-character/dnd-character.js
export const abilityModifier = (number) => {
    if (number < 3) { throw 'Ability scores must be at least 3'; }
    if (number > 18) { throw 'Ability scores can be at most 18'; }
    return Math.floor((number - 10) / 2);
};

export class Character {
    constructor() {
        this._strength = Character.rollAbility(); 
        this._dexterity = Character.rollAbility();
        this._constitution = Character.rollAbility();
        this._intelligence = Character.rollAbility();
        this._wisdom = Character.rollAbility();
        this._charisma = Character.rollAbility();
    }

    static rollAbility() {
        let diceDotsCount = 6;

        let roles = [Math.floor((Math.random() * diceDotsCount) + 1),
                     Math.floor((Math.random() * diceDotsCount) + 1),
                     Math.floor((Math.random() * diceDotsCount) + 1),
                     Math.floor((Math.random() * diceDotsCount) + 1)]

        roles.sort().shift();

        return roles.reduce((total, roll) => total + roll);
    } 

    get strength() {
        return this._strength;
    };
    get dexterity() {
        return this._dexterity;
    };
    get constitution() {
        return this._constitution;
    };
    get intelligence() {
        return this._intelligence;
    };
    get wisdom() {
        return this._wisdom;
    };   
    get charisma() {
        return this._charisma;
    };
    get hitpoints() { return 10 + abilityModifier(this._constitution) };
}

File: csharp/dnd-character/DndCharacter.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class DndCharacter
{
    public const int InitialHitPoint = 10;
    public const int MinScore = 1;
    public const int MaxScore = 6;
    public int Strength { get; }
    public int Dexterity { get; }
    public int Constitution { get; }
    public int Intelligence { get; }
    public int Wisdom { get; }
    public int Charisma { get; }
    public int Hitpoints { get; }

    public DndCharacter()
    {
        Strength = Ability();
        Dexterity = Ability();
        Constitution = Ability();
        Intelligence = Ability();
        Wisdom = Ability();
        Charisma = Ability();
        Hitpoints = InitialHitPoint + Modifier(Constitution);
    }

    public static int Modifier(int score)
    {
      decimal converstedScore = score;
      decimal value = Decimal.Divide(converstedScore - InitialHitPoint,  2m);
      MidpointRounding roundType = value > 0 ? MidpointRounding.ToZero : MidpointRounding.AwayFromZero;
      
      return (int)decimal.Round(value, roundType);
    }

    public static int Ability() 
    {
        var diceRole1 = new Random().Next(MinScore, MaxScore+1);
        var diceRole2 = new Random().Next(MinScore, MaxScore+1);
        var diceRole3 = new Random().Next(MinScore, MaxScore+1);
        var diceRole4 = new Random().Next(MinScore, MaxScore+1);
        
        var diceRoles = new []{ diceRole1, diceRole2, diceRole3, diceRole4 };
       
        return diceRoles.OrderByDescending(x => x).Take(diceRoles.Count()-1).Sum(x => x); 
    }

    public static DndCharacter Generate() =>new DndCharacter();
}
